c#/Traverse/TraverseFolders/FileSystemVisitor.cs

```diff
@@ -1,72 +1,71 @@
-public delegate bool ReturnedDelegate(string filename, string extention);
-
+public delegate bool FilterFunctionDelegate(string name);
+public delegate void Notify();
+public delegate void FindFolderDelegate(string filename,CustomEventArgs eventArgs);
 class FileSystemVisitor
 {
-    public FileSystemVisitor(string filename)
+    private readonly string _path;
+    private readonly FilterFunctionDelegate _filterFunc;
+
+    public event Notify OnStart;
+    public event Notify OnFinish;
+    public event FindFolderDelegate OnFileFound;
+    public event FindFolderDelegate OnFolderExclude;
+
+    public FileSystemVisitor(string path)
     {
-        foreach (string file in Explore(filename)) {
-            Console.WriteLine(file);
-        }
+        _path = path;
+        _filterFunc = (_) => true;
     }
 
-    public FileSystemVisitor(string filename, ReturnedDelegate delegat)
+    public FileSystemVisitor(string filename, FilterFunctionDelegate delegat) : this(filename)
     {
-        foreach (string file in Explore(filename,delegat)) {
-            Console.WriteLine(file);
-        }
+        _filterFunc = delegat;
     }
-    
-    public IEnumerable<string> Explore(string path, ReturnedDelegate delegat)
+
+    public IEnumerable<string> Explore(string path)
     {
+        OnStart.Invoke();
```

Comment:
This event will be called every time you call Explore method recursively, instead of being called only on the beginning of the process. The same thing will happen with OnFinish event.
To resolve this issue you can make this method private and create new parameterless public method which will call private one and call events before and after only once.

Example of the new method:
```
    public IEnumerable<string> Explore()
    {
        OnStart();

        foreach (var entry in Eplore(_rootDir))
        {
            yield return entry;
        }
        
        OnFinish();
    }
```